### PR TITLE
patron_type: implements overdue items restriction

### PR DESCRIPTION
**Pull request by zannkukai**
Implements the overdue_items_limit restriction : If a patron type
resource defines an 'overdue_items_limit', patrons with overdue items
can't execute any checkout operation if the limit is reached.
This commit also refactors the 'blocked_patron' restriction.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/items/api/circulation.py [hunk 1]
```diff
@@ -583,6 +592,11 @@ def extend_loan(self, current_loan, **kwargs):
     @add_action_parameters_and_flush_indexes
     def request(self, current_loan, **kwargs):
         """Request item for the user and create notifications."""
+        # Check if the 'checkout' action is allowed
```
Comment (lauren-d) [inline rero_ils/modules/items/api/circulation.py:595]: checkout or request?

**Review by lauren-d — APPROVED** [review]

File: rero_ils/modules/items/api/circulation.py (hunk 1 above)
Comment (zannkukai) [reply rero_ils/modules/items/api/circulation.py:595]: copy/paste and ICC problem ;-)
Corrected in next PR (avoiding rebase/cherry-pick....)

File: rero_ils/modules/items/api/circulation.py [hunk 2]
```diff
@@ -434,13 +435,21 @@ def compare_item_pickup_transaction_libraries(self, **kwargs):
     @add_action_parameters_and_flush_indexes
     def checkout(self, current_loan, **kwargs):
         """Checkout item to the user."""
+        # Check if the 'checkout' action is allowed
```
Comment (BadrAly) [inline rero_ils/modules/items/api/circulation.py:438]: I think it is better to move this check to the beginning of the method `add_action_parameters_and_flush_indexes`.  you do not need that we execute the logic of `add_action_parameters_and_flush_indexes` and then come to the function to find out that we are not authorized to do a checkout

File: rero_ils/modules/items/api/circulation.py [hunk 3]
```diff
@@ -583,6 +592,11 @@ def extend_loan(self, current_loan, **kwargs):
     @add_action_parameters_and_flush_indexes
     def request(self, current_loan, **kwargs):
         """Request item for the user and create notifications."""
+        # Check if the 'checkout' action is allowed
+        can, reasons = self.can(ItemCirculationAction.REQUEST, **kwargs)
```
Comment (BadrAly) [inline rero_ils/modules/items/api/circulation.py:596]: same remark here

File: rero_ils/modules/loans/api.py [hunk 4]
```diff
@@ -665,19 +665,35 @@ def get_due_soon_loans():
...
-    """Return all overdue loans."""
-    overdue_loans = []
-    results = current_circulation.loan_search_cls()\
-        .filter('term', state=LoanState.ITEM_ON_LOAN)\
-        .params(preserve_order=True)\
-        .sort({'_created': {'order': 'asc'}})\
+def get_overdue_loan_pids(patron_pid=None):
+    """Return all overdue loan pids optionally filtered for a patron pid.
+
+    :param patron_pid: the patron pid. If none, return all overdue loans.
```
Comment (BadrAly) [inline rero_ils/modules/loans/api.py:671]: the docs does not talk about the `range` filter

File: rero_ils/modules/loans/utils.py [hunk 5]
```diff
@@ -185,6 +177,12 @@ def can_be_requested(loan):
     location = Location.get_record_by_pid(loan.location_pid)
     if not location or not location.get('allow_request'):
         return False
+
+    # 4) Check patron is not blocked
+    patron = Patron.get_record_by_pid(loan.get('patron_pid'))
```
Comment (BadrAly) [inline rero_ils/modules/loans/utils.py:182]: patron = Patron.get_record_by_pid(loan.patron_pid)

File: rero_ils/modules/loans/utils.py [hunk 6]
```diff
@@ -185,6 +177,12 @@ def can_be_requested(loan):
     location = Location.get_record_by_pid(loan.location_pid)
     if not location or not location.get('allow_request'):
         return False
+
+    # 4) Check patron is not blocked
```
Comment (BadrAly) [inline rero_ils/modules/loans/utils.py:181]: The order of the execution is important here, I would move the patron blocking to be the first check. If a patron is blocked, I do not need to go and find a cipo (check2) or if he is allowed to request (check3) or even check1

**Review by BadrAly — CHANGES_REQUESTED** [review]

File: rero_ils/modules/loans/api.py (hunk 4 above)
Comment (zannkukai) [reply rero_ils/modules/loans/api.py:671]: the docstring talks about "overdue loan". An overdue_loan is a loan with ITEM_ON_LOAN state with a enddate <= systdate. This is the purpose of the 'range' filter. This filter is not a function argument.

File: rero_ils/modules/loans/utils.py (hunk 6 above)
Comment (zannkukai) [reply rero_ils/modules/loans/utils.py:181]: You're right. Done

File: rero_ils/modules/items/api/circulation.py (hunk 2 above)
Comment (zannkukai) [reply rero_ils/modules/items/api/circulation.py:438]: Right! However, I think it's even better to create a new decorator `@check_operation` called before `@add_action_parameters_and_flush_indexes`.

Comment (BadrAly) [reply rero_ils/modules/items/api/circulation.py]: Yes, even better

**Review by BadrAly — APPROVED** [review]

**Pull request merged**